packages/lib-auth/src/client.js

```diff
@@ -0,0 +1,318 @@
+   */
+  isSignedIn () {
+    return !!this._state.accessToken
+  }
+
+  /**
+   * Makes a request to exchange the refresh token for new token data.
+   *
+   * @async
+   * @private
+   * @returns {Promise} Resolves to the new access token.
+   */
+  async _refreshTokenData () {
+    const data = {
+      'client_id': this._config.clientAppID,
+      'grant_type': 'refresh_token',
+      'refresh_token': this._state.refreshToken
+    }
+    const response = await this._httpClient.post('/oauth/token', data)
+    this._handleNewTokenData(response.data)
+    return this._state.accessToken
+  }
+
+  /**
+   * Registers a new user and logs them in.
+   *
+   * @async
+   * @param {Object} user - The details of the new user account.
+   * @param {boolean} user.betaEmailCommunication - Whether the new user wants to opt in to beta emails.
+   * @param {string} user.creditedName - The new user's credited name (used in citations etc).
+   * @param {string} user.email - The new user's email address.
+   * @param {string} user.login - The new user's username.
+   * @param {boolean} user.globalEmailCommunication - Whether the new user wants to opt in to global emails.
+   * @param {string} user.password - The new user's password.
+   * @param {boolean} user.projectEmailCommunication - Whether the new user wants to opt in to project-related emails when registering via a specific project.
+   * @param {string} user.projectId - The project ID when registering via a specific project.
+   * @returns {Promise} Resolves to a new access token.
+   */
+  async register (user) {
+    if (this.isSignedIn()) {
+      await this.signOut()
+    }
+
+    const csrfToken = await this._getCSRFToken()
+    const data = {
+      'authenticity_token': csrfToken,
+      user: {
+        'beta_email_communication': user.betaEmailCommunication,
+        'credited_name': user.creditedName,
+        email: user.email,
+        'global_email_communication': user.globalEmailCommunication,
+        login: user.login,
+        password: user.password,
+        'project_email_communication': user.projectEmailCommunication,
+        'project_id': user.projectId
+      }
+    }
+    const registerResponse = await this._httpClient.post('/users', data)
+
+    // After the user is created, we need to sign in to actually get an access
+    // token. `panoptes-javascript-client` manages to do it without this step;
+    // there's no `set-cookie` header on the response that I can see, so I'm not
+    // actually sure how.
+    const credentials = {
+      login: user.login,
+      password: user.password,
+    }
+    return this.signIn(credentials)
+  }
+
+  /**
+   * Resets the client state to its initial state
+   *
+   * @private
+   * @returns {void}
+   */
+  _resetState () {
+    this._state = {
+      accessToken: '',
+      accessTokenCreatedAt: NaN,
+      accessTokenExpiresAt: NaN,
+      accessTokenExpiresIn: NaN,
+      refreshToken: '',
+      scope: ''
+    }
+  }
```

Comment:
Is there a better way of storing the tokens than making them publicly available in `client._state`?